assumeutxo: snapshot initialization

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11) (parent PR: https://github.com/bitcoin/bitcoin/pull/15606)

---

Half of the replacement for #24232. The original PR grew larger than expected throughout the review process.

This change adds the ability to initialize a snapshot-based chainstate during init if one is detected on disk. This is of course unused as of now (aside from in unittests) given that we haven't yet enabled actually loading snapshots. 

Don't be scared! There are some big move-only commits in here.

Accompanying changes include: 

- moving the snapshot coinsdb directory from being called `chainstate_[base blockhash]` to `chainstate_snapshot`, since we only support one snapshot in use at a time. This simplifies some logic, but it necessitates writing that base blockhash out to a file within the coinsdb dir. See [discussion here](https://github.com/bitcoin/bitcoin/pull/24232#discussion_r832762880).
- adding a simple fix in `FlushBlockFile()` that avoids a crash when attemping to flush to disk before `LoadBlockIndexDB()` is called, which happens when calling `MaybeRebalanceCaches()` during multiple chainstate init.
- improving the unittest to allow testing with on-disk chainstates - necessary to test a simulated restart and re-initialization.

jamesob committed:
validation: remove snapshot datadirs upon validation failure

If a UTXO snapshot fails to validate, don't leave the resulting datadir
on disk as this will confuse initialization on next startup and we'll
get an assertion error.

jamesob committed:
db: add StoragePath to CDBWrapper/CCoinsViewDB

This is used in subsequent commits. It allows us to clean up UTXO
snapshot chainstate after background validation completes.

jamesob committed:
blockmanager: avoid undefined behavior during FlushBlockFile

If we call FlushBlockFile() without having intitialized the block index
with LoadBlockIndexDB(), we may be indexing into an empty vector.

Specifically this is an issue when we call MaybeRebalanceCaches() during
chainstate init before the block index has been loaded, which calls
FlushBlockFile().

Also add an assert to avoid undefined behavior.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob committed:
move-only: test: make snapshot chainstate setup reusable

For use in next commit.

Most easily reviewed with
`--color-moved=dimmed_zebra --color-moved-ws=allow-indentation-change`.

jamesob committed:
test: add testcases for snapshot initialization

jamesob committed:
test: add reset_chainstate parameter for snapshot unittests

This CreateAndActivateUTXOSnapshot parameter is necessary once we
perform snapshot completion within ABC, since the existing UpdateTip
test will fail because the IBD chain that has generated the snapshot
will exceed the base of the snapshot.

Being able to test snapshots being loaded into a mostly-uninitialized
datadir allows for more realistic unittest scenarios.

jamesob committed:
test: allow on-disk coins and block tree dbs in tests

Used when testing cleanup of on-disk chainstate data for snapshot
testcases. Also necessary for simulating node restart in .cpp tests.

jamesob committed:
validation: rename snapshot chainstate dir

This changes the snapshot's leveldb chainstate dir name from
`chainstate_[blockhash]` to `chainstate_snapshot`. This simplifies
later logic that loads snapshot data, and enforces the limitation
of a single snapshot at any given time.

Since we still need to persis the blockhash of the base block, we
write that out to a file (`chainstate_snapshot/base_blockhash`) for
later use during initialization, so that we can reinitialize the
snapshot chainstate.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob committed:
init: add utxo snapshot detection

Add functionality for activating a snapshot-based chainstate if one is
detected on-disk.

Also cautiously initialize chainstate cache usages so that we don't
somehow blow past our cache allowances during initialization, then
rebalance at the end of init.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob committed:
test: move-only-ish: factor out LoadVerifyActivateChainstate()

in TestingSetup(). This is used in the following commit to test
reinitializing chainstates after snapshot validation and cleanup.

Best reviewed with `git diff --color-moved=dimmed-zebra`.

jamesob committed:
add utilities for deleting on-disk leveldb data

Used in later commits to remove leveldb directories for
- invalid snapshot chainstates, and
- background-vaildation chainstates that have finished serving their
  purpose.

jamesob committed:
validation: add ResetChainstates()

Necessary for the following test commit.

ryanofsky commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -328,6 +354,31 @@ struct SnapshotTestSetup : TestChain100Setup {
+                ::Params(),
+                static_cast<int64_t(*)()>(::GetTime),
+            };
+            m_node.chainman.reset(new ChainstateManager(chainman_opts));
```
In commit "test: add testcases for snapshot initialization" (15da925f2951eaadf81aea9d5b67ec1646581d30)

Might be a little more robust to destroy the old chainman before creating a new one. I.e. add another blank `chainman.reset()` call before this.

jamesob replied:
Fixed, thanks.

ryanofsky reviewed (approved):
Code review ACK 15da925f2951eaadf81aea9d5b67ec1646581d30. All commits here except the last two were directly cherry-picked from previously reviewed #24232.

ariard commented on src/validation.cpp:
```diff
@@ -4720,6 +4708,46 @@ const AssumeutxoData* ExpectedAssumeutxo(
     return nullptr;
 }
 
+static bool DeleteCoinsDBFromDisk(const fs::path db_path, bool is_snapshot)
```
note for reviewers: for now `is_snapshot` is always true as there is a single usage. This is not the case anymore with the remaining commits of the #24232 patchset.

ariard commented on src/validation.h:
```diff
@@ -990,6 +984,15 @@ class ChainstateManager
+
+    void ResetChainstates() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    CChainState& ActivateExistingSnapshot(CTxMemPool* mempool, uint256 base_blockhash)
```
even if name is explicit, could be documented "Switch active chainstate to snapshot one" in case of future refactor to check function correctness (and `DetectSnapshotChainstate` could say it activates the snapshot if any)

jamesob replied:
Done, thanks.

ariard commented on src/node/utxo_snapshot.h:
```diff
@@ -33,6 +39,30 @@ class SnapshotMetadata
+
+//! The file in the snapshot chainstate dir which stores the base blockhash. This
+//! is needed to reconstruct snapshot chainstates on init.
+const fs::path SNAPSHOT_BLOCKHASH_FILENAME{"base_blockhash"};
```
comment can be extended "There should be a single snapshot at any given time"

jamesob replied:
Done.

ariard commented on src/validation.h:
```diff
@@ -990,6 +984,15 @@ class ChainstateManager
+    //! snapshot that is in the process of being validated.
+    bool DetectSnapshotChainstate(CTxMemPool* mempool) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    void ResetChainstates() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
note to reviewers: this method is used in the remaining #24232 patchset.

ariard reviewed (commented):
Code Review ACK 15da925f, my comments on #24232 have been addressed.

ariard commented:
@dergoegge I think you could be interested to review the assumeutxo work. In case of future hypothetical integration of libutreexo in Core, I believe there would be intersecting code paths and integration. It might also prepare for future [synergies](https://github.com/jamesob/assumeutxo-docs/tree/master/proposal#how-will-this-work-with-accumulators-uhs-or-utreexo-if-those-things-come-around) between utreexo and assumeutxo.

dongcarl commented:
Planning on diving deeply into this, thanks for splitting it up @jamesob!

Questions from going over the commits:

Q1:  Is the intended layout:
- `<datadir>`
  - `chainstate`
    - `*.ldb`
    - `*.log`
    - `CURRENT`
    - `LOCK`
    - `MANIFEST-*`
  - `chainstate_snapshot`
    - `base_blockhash`
    - `*.ldb`
    - `*.log`
    - `CURRENT`
    - `LOCK`
    - `MANIFEST-*` 

Q2: The last 6 commits seem to be focused on unit tests, but it seems to me that the functionality being added here is better tested with functional tests. Is there something we need to test here that a functional test won't be able to do?

dongcarl commented on src/validation.cpp:
```diff
@@ -4708,6 +4708,46 @@ const AssumeutxoData* ExpectedAssumeutxo(
+        fs::path base_blockhash_path = db_path / node::SNAPSHOT_BLOCKHASH_FILENAME;
+
+        if (fs::exists(base_blockhash_path)) {
+            bool removed = fs::remove(base_blockhash_path);
+            if (!removed) {
+                LogPrintf("[snapshot] failed to remove file %s\n",
+                          fs::PathToString(base_blockhash_path));
+            }
+        } else {
+            LogPrintf("[snapshot] snapshot chainstate dir being removed lacks %s file\n",
+                    fs::PathToString(node::SNAPSHOT_BLOCKHASH_FILENAME));
+        }
+    }
+
+    std::string path_str = fs::PathToString(db_path);
+    LogPrintf("Removing leveldb dir at %s\n", path_str);
+
+    // We have to destruct before this call leveldb::DB in order to release the db
+    // lock, otherwise `DestroyDB` will fail. See `leveldb::~DBImpl()`.
+    const bool destroyed = dbwrapper::DestroyDB(path_str, {}).ok();
```
Note to other reviewers: the call to `DestroyDB` will actually remove the entire directory, including the blockhash file if it still existed since we place the blockhash file inside what leveldb considers to be its directory, but it's probably better to remove it ourselves just so we can print nice errors.

https://github.com/bitcoin/bitcoin/blob/de3c46c93818c6e4000175bcdbf7dd332f54768d/src/leveldb/db/db_impl.cc#L1550

ariard commented:
This PR needs rebase.

jamesob committed:
doc: add note about snapshot chainstate init

jamesob commented on src/node/chainstate.cpp:
```diff
@@ -31,10 +31,15 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
+    chainman.InitializeChainstate(options.mempool);
+
+    // Load a chain created from a UTXO snapshot, if any exist.
+    chainman.DetectSnapshotChainstate(options.mempool);
```
This wasn't documented, but I've just added some related doc in this PR. There are more doc updates, of course, in the follow-on PR (#25740).

jamesob commented:
Hi all, sorry for the delay. My spirit has been weak the last few months, and I took two weeks off to try and shake the burnout. 

I've addressed all feedback listed here and rebased.

@dongcarl asks:

> Q1: Is the intended layout:

Yep, it is.

> Q2: The last 6 commits seem to be focused on unit tests, but it seems to me that the functionality being added here is better tested with functional tests. Is there something we need to test here that a functional test won't be able to do?

We can't have functional tests to exercise this behavior until we add `loadtxoutset` to the RPC interface, which will effectively reveal assumeutxo to end users before all the necessary changes have been made to make it usable. Plus, I think the unittests allow us to more easily express granular testcases and assertions.

naumenkogs commented on doc/design/assumeutxo.md:
```diff
@@ -77,7 +77,9 @@ original chainstate remains in use as active.
 chainstate and a sync to tip begins. A new chainstate directory is created in the
 datadir for the snapshot chainstate called
-`chainstate_[SHA256 blockhash of snapshot base block]`.
+`chainstate_[SHA256 blockhash of snapshot base block]`. When this directory is present
```
isn't this outdated now after d59db4ff4b585d2b86b68ea0db206ec1c0cc5022?

jamesob replied:
Oops, good catch.

naumenkogs commented:
utACK 1430518eeced979449c3b031bea02fd707b26d1a.

MarcoFalke commented on src/dbwrapper.h:
```diff
@@ -219,6 +223,12 @@ class CDBWrapper
 
     std::vector<unsigned char> CreateObfuscateKey() const;
 
+    //! path to filesystem storage
+    const fs::path m_path;
+
+    //! whether or not the database resides in memory
+    bool m_is_memory;
+
```
nit in 75bb0c54b383e434695bf6bc1d60ef5bcd83cd70:

Can combine both into

```cpp
    //! path to filesystem storage or std::nullopt if database resides in memory
    std::optional<fs::path> m_path;

jamesob replied:
I think @ryanofsky brought this up earlier; while I agree it would be a good change, it will broaden the scope of this PR quite a bit because all uses of CDBWrapper will have to change, since `path` is currently a required argument. PR is already large, so this could be a follow-up.

MarcoFalke commented on src/node/utxo_snapshot.cpp:
```diff
@@ -0,0 +1,79 @@
+    const fs::path write_to = *chaindir / node::SNAPSHOT_BLOCKHASH_FILENAME;
+
+    FILE* file{fsbridge::fopen(write_to, "wb")};
+    CAutoFile afile{file, SER_DISK, CLIENT_VERSION};
```
nit in d59db4ff4b585d2b86b68ea0db206ec1c0cc5022:

```diff
diff --git a/src/node/utxo_snapshot.cpp b/src/node/utxo_snapshot.cpp
index 9c9947df95..2c5885198b 100644
--- a/src/node/utxo_snapshot.cpp
+++ b/src/node/utxo_snapshot.cpp
@@ -25,7 +25,7 @@ bool WriteSnapshotBaseBlockhash(CChainState& snapshot_chainstate)
     const fs::path write_to = *chaindir / node::SNAPSHOT_BLOCKHASH_FILENAME;
 
     FILE* file{fsbridge::fopen(write_to, "wb")};
-    CAutoFile afile{file, SER_DISK, CLIENT_VERSION};
+    AutoFile afile{file};
     if (afile.IsNull()) {
         LogPrintf("[snapshot] failed to open base blockhash file for writing: %s\n",
                   fs::PathToString(write_to));
@@ -60,7 +60,7 @@ std::optional<uint256> ReadSnapshotBaseBlockhash(fs::path chaindir)
 
     uint256 base_blockhash;
     FILE* file{fsbridge::fopen(read_from, "rb")};
-    CAutoFile afile{file, SER_DISK, CLIENT_VERSION};
+    AutoFile afile{file};
     if (afile.IsNull()) {
         LogPrintf("[snapshot] failed to open base blockhash file for reading: %s\n",
             read_from_str);

ariard replied:
I think it could be valuable to document `CAutoFile` in `streams.h`, there is no document to explain the diff compared to `AutoFile`. From my understanding, it's a wrapper to handle file pointer, without having to bother with serialization type and versioning.

MarcoFalke commented on src/node/utxo_snapshot.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    const std::optional<fs::path> chaindir = snapshot_chainstate.CoinsDB().StoragePath();
+    assert(chaindir); // Sanity check that chainstate isn't in-memory.
+    const fs::path write_to = *chaindir / node::SNAPSHOT_BLOCKHASH_FILENAME;
```
question in https://github.com/bitcoin/bitcoin/commit/d59db4ff4b585d2b86b68ea0db206ec1c0cc5022: why can't the blockhash be read from the metadata in the dump?

jamesob replied:
By "dump" do you mean the UTXO snapshot? If so, I don't think we can expect users to keep the snapshot on-hand after they've loaded it; it's essentially unnecessary at that point, and multiple GBs.

MarcoFalke commented on src/test/util/chainstate.h:
```diff
@@ -48,7 +62,39 @@ CreateAndActivateUTXOSnapshot(node::NodeContext& node, const fs::path root, F ma
+            node.chainman->ResetChainstates();
+            node.chainman->InitializeChainstate(node.mempool.get());
+            CChainState& chain = node.chainman->ActiveChainstate();
+            BOOST_CHECK(chain.LoadGenesisBlock());
```
```suggestion
            Assert(chain.LoadGenesisBlock());
```

might be better to not continue execution when this fails?

MarcoFalke commented on src/test/util/chainstate.h:
```diff
@@ -48,7 +62,39 @@ CreateAndActivateUTXOSnapshot(node::NodeContext& node, const fs::path root, F ma
+        if (!node.chainman->ActiveChainstate().ActivateBestChain(state)) {
+            throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", state.ToString()));
+        }
+        BOOST_CHECK_EQUAL(
```
```suggestion
        Assert(
```

maflcko reviewed (commented):
left a question. Also, this no longer compiles because boost was removed from libtest_util

naumenkogs commented:
utACK bf9597606166323158bbf631137b82d41f39334f

ariard reviewed (approved):
Code Review ACK bf9597606

Changes since last ACK:
- `CChainState` -> `Chainstate`, after #24513
- replacing `CAutoFile` usage in `utxo_snapshot.cpp` by `AutoFile`
- commenting `SNAPSHOT_BLOCKHASH_FILENAME`
- a new informative log L116 in `src/node/chainstate.cpp` 
- replacing `InitializeChainstate` by `ActivateExistingSnapshot` in tests
- dropping the optional `snapshot_blockash` argument from `InitializeChainstate` and adding asserts
- commenting `ActivateExistingSnapshot`
- comment, lock simplification and minor code changes in unit tests

ryanofsky reviewed (approved):
Code review ACK bf9597606166323158bbf631137b82d41f39334f. Changes since last review: rebasing, switching from CAutoFile to AutoFile, adding comments, switching from BOOST_CHECK to Assert in test util, using chainman.GetMutex() in tests, destroying one ChainstateManager before creating a new one in tests

fjahr reviewed (commented):
utACK bf9597606166323158bbf631137b82d41f39334f

jamesob commented:
I'm counting 4 ACKs here - what more is needed to merge?

aureleoules reviewed (commented):
LGTM, I also verified that all commits compile and pass unit tests.
Left some minor/nit comments.

jamesob commented:
I'm going to leave the nits as-is to avoid ACK churn unless maints feel they are blocking merge. I'll be very quick to review any followups that @aureleoules or @fjahr would like to do towards addressing them.

aureleoules reviewed (approved):
ACK bf9597606166323158bbf631137b82d41f39334f

andrewtoth commented on src/validation.cpp:
```diff
@@ -4767,6 +4767,46 @@ const AssumeutxoData* ExpectedAssumeutxo(
+        if (fs::exists(base_blockhash_path)) {
+            bool removed = fs::remove(base_blockhash_path);
+            if (!removed) {
+                LogPrintf("[snapshot] failed to remove file %s\n",
```
I don't think this code is the intended behavior. This is redundant to the previous check 3 lines up. If `fs::remove` without a second parameter returns `false` it is only because the file does not exist. If it actually fails to remove an existing file it will throw an exception (see https://en.cppreference.com/w/cpp/filesystem/remove).

If intended behavior is to only log error and continue `fs::remove` should be wrapped in a try/catch block or passed a `std::error_code`.
